lib/plug/debugger/highlighter.ex

```diff
@@ -0,0 +1,312 @@
+  end
+
+  @doc """
+  Highlights the arguments for a stack frame.
+
+  Receives a list of elixir values and outputs a list of HTML strings.
+  No need to care about splitting at linebreaks.
+  """
+  def highlight_args(args) do
+    if Code.ensure_loaded?(ElixirLexer) do
+      Enum.map(args, &highlight_arg/1)
+    else
+      plain_args(args)
+    end
+  end
+
+  @doc """
+  Highlights a sequence of lines of source code.
+
+  Returns a list of highlighted lines in the format expected by `Plug.Debugger`.
+  Must split tokens at linebreaks.
+  """
+  def highlight_snippet(lines, file) do
+    case lexer_for_file(file) do
+      _ -> lines
+      ElixirLexer -> highlight_snippet_with_lexer(lines, ElixirLexer)
+      _other -> plain_lines(lines)
+    end
+  end
+
+  # ----------------------------
+  # Private functions
+  # ----------------------------
+
+  # Highlighting function arguments:
+  # --------------------------------
+
+  # Escape arguments not meant to be highlighted,
+  # so that they can be inserted in the HTML template
+  defp plain_args(args) do
+    Enum.map(args, fn arg -> arg |> inspect() |> escape_string() end)
+  end
+
+  # Highlight a list of arguments
+  defp highlight_arg(arg) do
+    # Makeup takes care of the escaping
+    arg |> inspect() |> Makeup.highlight_inner_html()
+  end
+
+  # Highlight source lines
+  # -----------------------------
+
+  # Evaluates the maximum number of digits in a list of integers
+  defp max_digits([]), do: 0
+
+  defp max_digits(numbers) do
+    numbers
+    |> Enum.max()
+    |> :math.log10()
+    |> :math.ceil()
+    |> round()
+  end
+
+  # Converts the line number into a string and pads it with wthitspace
+  # so that it occupied the same space as the maximum number of digits
+  # Number `5` might become `" 5"` or `"  5"`, depending
+  # on the maximum line number
+  defp pad_line_number(nr, length) do
+    String.pad_leading(to_string(nr), length)
+  end
+
+  defp pad_lines(lines) do
+    line_numbers = Enum.map(lines, fn {nr, _, _} -> nr end)
+    length = max_digits(line_numbers)
+    Enum.map(lines, fn {nr, line, highlight} ->
+      {pad_line_number(nr, length), line, highlight}
+    end)
+  end
+
+  # Stolen from `Plug.Debugger`
+  defp escape_string(string) do
+    string |> to_string() |> String.trim_trailing() |> Plug.HTML.html_escape()
+  end
+
+  defp escape_line({nr, line, highlighted}) do
+    {nr, escape_string(line), highlighted}
+  end
+
+  defp escape_lines(lines) do
+    Enum.map(lines, &escape_line/1)
+  end
+
+  defp plain_lines(lines) do
+    lines
+    |> pad_lines()
+    |> escape_lines()
+  end
+
+  # Currently only an elixir lexer is supported
+  defp lexer_for_file(file) do
+    cond do
+      String.ends_with?(file, ".ex") -> ElixirLexer
+      String.ends_with?(file, ".exs") -> ElixirLexer
+      :otherwise -> nil
+    end
+  end
+
+  defp highlight_snippet_with_lexer(lines, lexer) do
+    if Code.ensure_loaded?(lexer) do
```

Comment:
It feels like a lot of the work here is undoing the work done in `Plug.Debugger`. So maybe reading the file here will streamline this a lot.